src/lib.rs

```diff
@@ -483,16 +522,105 @@ impl Bump {
 
         Ok(Bump {
             current_chunk_footer: Cell::new(chunk_footer),
+            allocation_limit: Cell::new(None),
         })
     }
 
+    /// The allocation limit for this arena in bytes.
+    ///
+    /// ## Example
+    ///
+    /// ```
+    /// let bump = bumpalo::Bump::with_capacity(0);
+    ///
+    /// assert_eq!(bump.allocation_limit(), None);
+    ///
+    /// bump.set_allocation_limit(6);
+    ///
+    /// assert_eq!(bump.allocation_limit(), Some(6));
+    ///
+    /// bump.remove_allocation_limit();
+    ///
+    /// assert_eq!(bump.allocation_limit(), None);
+    /// ```
+    pub fn allocation_limit(&self) -> Option<usize> {
+        self.allocation_limit.get()
+    }
+
+    /// Set the allocation limit in bytes for this arena.
+    ///
+    /// Changing the limit for a `Bump` while it has live allocations does
+    /// nothing until a new allocation is attempted, at which point it may fail
+    /// due to the new limit.
+    ///
+    /// ## Example
+    ///
+    /// ```
+    /// let bump = bumpalo::Bump::with_capacity(0);
+    ///
+    /// bump.set_allocation_limit(0);
+    ///
+    /// assert!(bump.try_alloc(5).is_err());
+    /// ```
+    pub fn set_allocation_limit(&self, new_limit: usize) {
+        self.allocation_limit.set(Some(new_limit))
+    }
+
+    /// Remove the allocation limit in bytes for this arena.
+    ///
+    /// ## Example
+    ///
+    /// ```
+    /// let bump = bumpalo::Bump::with_capacity(0);
+    ///
+    /// bump.set_allocation_limit(0);
+    ///
+    /// assert!(bump.try_alloc(5).is_err());
+    ///
+    /// bump.remove_allocation_limit();
+    ///
+    /// assert!(bump.try_alloc(5).is_ok());
+    /// ```
+    pub fn remove_allocation_limit(&self) {
+        self.allocation_limit.set(None)
+    }
+
+    /// How much headroom an arena has before it hits its allocation
+    /// limit.
+    #[inline(always)]
+    fn allocation_limit_remaining(&self) -> Option<usize> {
+        self.allocation_limit
+            .get()
+            .map(|allocation_limit| {
+                let allocated_bytes = self.allocated_bytes();
```

Comment:
`allocated_bytes` has to walk over the chunks linked list to sum up all the chunk sizes, and I don't think we really want to introduce that kind of walk to new-chunk allocation. It is an `O(n)` walk, that we will do `n` times for each new chunk allocation, so it will lead to a classic `O(n^2)` behavior.

Instead, I think the `ChunkFooter` needs to start recording the total byte size of allocated chunks thus far, that is:

* the canonical empty chunk will have `allocated_bytes = 0`,
* chunk A, whose `prev = canonical empty chunk`, will have `allocated_bytes = 0 + size(A)`
* chunk B, whose `prev = A`, will have `allocated_bytes = A.allocated_bytes + size(B)`,
* etc...

Then we can have `allocated_bytes` return the current footer's `allocated_bytes` instead of walking the chunk list, and then we will be able to safely use `self.allocated_bytes()` here without introducing `O(n^2)` behavior.

The one catch is that we will need to reset our chunk's `allocated_bytes` in the `Bump::reset` method to keep it in sync with how many bytes are allocated with the `Bump`.

Reply:
I implemented this in the latest commit, I think how I used `new_size_without_footer` to add to the `allocated_bytes` should preserve the old behavior but I'm not sure.